torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -935,27 +935,34 @@ void ProcessGroupNCCL::ncclCommWatchdog() {
   }
 }
 
-void ProcessGroupNCCL::logWorkStart(WorkNCCL& work) {
-  if (work.startTraceUpdated_)
+void ProcessGroupNCCL::logWorkStart(WorkNCCL& work, bool emitDesyncInfo) {
```

Comment:
ProcessGroupNCCL knows exactly when a collective error/timeout occurs. Should we add callbacks for those too?

Reply:
are you referring to watchdog functionality? i think the point of these hooks is to start unbundling some of the watchdog functionality to another more reliable service, hence, not building further on top of that.  (apart from using it to call the hooks)

Reply:
I meant the watchdog could just call the hooks when it detects a failure/timeout and all the heavy lifting around dealing with failures/timeouts could happen elsewhere as you suggest.

For example, the current start and end hooks can be used to detect a timeout but it won't tell you about errors. The watchdog uses [ncclCommGetAsyncError](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclcommgetasyncerror) to detect errors and it would be good to have visibility into those as well.

Reply:
There are two more events that are worth surfacing.

Collective enqueueing, which would help detect some classes of deadlocks.
Collective error, this is really tricky since when errors happen we might have a deadlocked python but still important.